Failure when parameter for the config file is first

While working on my xunit-reporter Quarkus based bot, I added customization through a config file.
So I changed my method from:

`void onWorkflow(@WorkflowRun.Completed GHEventPayload.WorkflowRun workflowRunPayload) throws IOException`

to:

`void onWorkflow(@ConfigFile(CONFIG_FILE_NAME) XUnitReporterConfigFile config, @WorkflowRun.Completed GHEventPayload.WorkflowRun workflowRunPayload) throws IOException`

But quarkus:dev now reports errors:

```
2021-04-22 11:57:40,170 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build f
ailed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedReso
lutionException: Unsatisfied dependency for type org.kohsuke.github.GHEventPayload$WorkflowRun and qualifiers [@Default]
        - java member: me.xunitreporter.XUnitBot_Multiplexer#onWorkflow_5faf52dd037aecb35cbd62097faa01b3382161b8()
        - declared on ObserverInfo [beanClass=me.xunitreporter.XUnitBot_Multiplexer, priority=2500, isAsync=true, reception=ALWAYS, transactionPhase=IN_PROGRESS, observedType=io.quarki
verse.githubapp.runtime.ConfigFileReader, qualifiers=[@Completed]]
        at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1081)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:255)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:129)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:419)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:920)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.kohsuke.github.GHEventPayload$WorkflowRun and qualifiers [@Default]
        - java member: me.xunitreporter.XUnitBot_Multiplexer#onWorkflow_5faf52dd037aecb35cbd62097faa01b3382161b8()
        - declared on ObserverInfo [beanClass=me.xunitreporter.XUnitBot_Multiplexer, priority=2500, isAsync=true, reception=ALWAYS, transactionPhase=IN_PROGRESS, observedType=io.quarki
verse.githubapp.runtime.ConfigFileReader, qualifiers=[@Completed]]
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:484)
        at io.quarkus.arc.processor.ObserverInfo.init(ObserverInfo.java:263)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:250)
        ... 12 more

        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:396)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:269)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:66)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:79)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:378)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:56)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:127)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:84)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:144)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedReso
lutionException: Unsatisfied dependency for type org.kohsuke.github.GHEventPayload$WorkflowRun and qualifiers [@Default]
        - java member: me.xunitreporter.XUnitBot_Multiplexer#onWorkflow_5faf52dd037aecb35cbd62097faa01b3382161b8()
        - declared on ObserverInfo [beanClass=me.xunitreporter.XUnitBot_Multiplexer, priority=2500, isAsync=true, reception=ALWAYS, transactionPhase=IN_PROGRESS, observedType=io.quarki
verse.githubapp.runtime.ConfigFileReader, qualifiers=[@Completed]]
        at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1081)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:255)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:129)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:419)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:920)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.kohsuke.github.GHEventPayload$WorkflowRun and qualifiers [@Default]
        - java member: me.xunitreporter.XUnitBot_Multiplexer#onWorkflow_5faf52dd037aecb35cbd62097faa01b3382161b8()
        - declared on ObserverInfo [beanClass=me.xunitreporter.XUnitBot_Multiplexer, priority=2500, isAsync=true, reception=ALWAYS, transactionPhase=IN_PROGRESS, observedType=io.quarki
verse.githubapp.runtime.ConfigFileReader, qualifiers=[@Completed]]
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:484)
        at io.quarkus.arc.processor.ObserverInfo.init(ObserverInfo.java:263)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:250)
        ... 12 more

        at io.quarkus.builder.Execution.run(Execution.java:116)
        at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:153)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:394)
        ... 9 more
Caused by: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.kohsuke.github.GHEventPayload$Wo
rkflowRun and qualifiers [@Default]
        - java member: me.xunitreporter.XUnitBot_Multiplexer#onWorkflow_5faf52dd037aecb35cbd62097faa01b3382161b8()
        - declared on ObserverInfo [beanClass=me.xunitreporter.XUnitBot_Multiplexer, priority=2500, isAsync=true, reception=ALWAYS, transactionPhase=IN_PROGRESS, observedType=io.quarki
verse.githubapp.runtime.ConfigFileReader, qualifiers=[@Completed]]
        at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1081)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:255)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:129)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:419)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:920)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.kohsuke.github.GHEventPayload$WorkflowRun and qualifiers [@Default]
        - java member: me.xunitreporter.XUnitBot_Multiplexer#onWorkflow_5faf52dd037aecb35cbd62097faa01b3382161b8()
        - declared on ObserverInfo [beanClass=me.xunitreporter.XUnitBot_Multiplexer, priority=2500, isAsync=true, reception=ALWAYS, transactionPhase=IN_PROGRESS, observedType=io.quarki
verse.githubapp.runtime.ConfigFileReader, qualifiers=[@Completed]]
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:484)
        at io.quarkus.arc.processor.ObserverInfo.init(ObserverInfo.java:263)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:250)
        ... 12 more

2021-04-22 11:57:40,170 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start hot replacement endpoint to recover from previous Quarkus startup failure
```

Having the config file parameter moved after the event payload fixed the issue

--- a/deployment/src/main/java/io/quarkiverse/githubapp/deployment/GitHubAppProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/githubapp/deployment/GitHubAppProcessor.java
@@ -595,13 +595,37 @@ class GitHubAppProcessor {
                     }
                 }
 
+                // copy annotations except for @ConfigFile
                 for (short i = 0; i < originalMethodParameterTypes.size(); i++) {
                     List<AnnotationInstance> parameterAnnotations = originalMethodParameterAnnotationMapping.getOrDefault(i,
                             Collections.emptyList());
-                    AnnotatedElement generatedParameterAnnotations = methodCreator.getParameterAnnotations(i);
+                    if (parameterAnnotations.isEmpty()) {
+                        continue;
+                    }
+
+                    // @ConfigFile elements are not in the mapping
+                    Short generatedParameterIndex = parameterMapping.get(i);
+                    if (generatedParameterIndex == null) {
+                        continue;
+                    }
+
+                    AnnotatedElement generatedParameterAnnotations = methodCreator
+                            .getParameterAnnotations(generatedParameterIndex);
                     if (parameterAnnotations.stream().anyMatch(ai -> ai.name().equals(eventSubscriberInstance.name()))) {
                         generatedParameterAnnotations.addAnnotation(DotNames.OBSERVES_ASYNC.toString());
                         generatedParameterAnnotations.addAnnotation(eventSubscriberInstance);
+                    } else {
+                        for (AnnotationInstance annotationInstance : parameterAnnotations) {
+                            generatedParameterAnnotations.addAnnotation(annotationInstance);
+                        }
+                    }
+                }
+
+                // generate the code of the method
+                for (short i = 0; i < originalMethodParameterTypes.size(); i++) {
+                    List<AnnotationInstance> parameterAnnotations = originalMethodParameterAnnotationMapping.getOrDefault(i,
+                            Collections.emptyList());
+                    if (parameterAnnotations.stream().anyMatch(ai -> ai.name().equals(eventSubscriberInstance.name()))) {
                         parameterValues[i] = methodCreator.getMethodParam(parameterMapping.get(i));
                     } else if (parameterAnnotations.stream().anyMatch(ai -> ai.name().equals(CONFIG_FILE))) {
                         AnnotationInstance configFileAnnotationInstance = parameterAnnotations.stream()
@@ -621,9 +645,6 @@ class GitHubAppProcessor {
                                 methodCreator.loadClass(configObjectType));
                         parameterValues[i] = methodCreator.checkCast(configObject, configObjectType);
                     } else {
-                        for (AnnotationInstance annotationInstance : parameterAnnotations) {
-                            generatedParameterAnnotations.addAnnotation(annotationInstance);
-                        }
                         parameterValues[i] = methodCreator.getMethodParam(parameterMapping.get(i));
                     }
                 }

--- a/deployment/src/test/java/io/quarkiverse/githubapp/deployment/ConfigFileOrderTest.java
+++ b/deployment/src/test/java/io/quarkiverse/githubapp/deployment/ConfigFileOrderTest.java
@@ -1,0 +1,31 @@
+package io.quarkiverse.githubapp.deployment;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.kohsuke.github.GHEventPayload;
+
+import io.quarkiverse.githubapp.ConfigFile;
+import io.quarkiverse.githubapp.event.Label;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ConfigFileOrderTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClass(ListeningClass.class))
+            .withConfigurationResource("application.properties");
+
+    @Test
+    public void testConfigFileOrder() {
+    }
+
+    static class ListeningClass {
+
+        void createLabel(@ConfigFile("my-config-file.txt") String configuration,
+                @Label.Created GHEventPayload.Label labelPayload) {
+        }
+    }
+}

